File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {

    private boolean isValid(int x, int y, int a, int b, char[][] grid, boolean[][] visited){
        if(x<0 || x>=a)
            return false;
        if(y<0 || y>=b)
            return false;

        if(grid[x][y] == '0' || visited[x][y])
            return false;
        
        return true;
    }

    private void solve(char[][] grid, boolean[][] visited, int x, int y){
        if(!isValid(x, y, grid.length, grid[0].length, grid, visited))
            return;

        visited[x][y] = true;

        solve(grid, visited, x+1, y);
        solve(grid, visited, x-1, y);
        solve(grid, visited, x, y+1);
        solve(grid, visited, x, y-1);
    }

    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int ans = 0;

        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                if(!visited[i][j] && grid[i][j] == '1'){
                    ans++;
                    solve(grid, visited, i, j);
                }
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.cpp
class Solution {
   public:
    int dir[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    bool isValid(int x, int y, int m, int n) {
        if (x < 0 || x >= m || y < 0 || y >= n) {
            return false;
        }

        return true;
    }

    void dfs(int x, int y, vector<vector<int>>& grid, vector<vector<bool>>& vis) {
        if (vis[x][y]) return;

        vis[x][y] = true;

        for (int i = 0; i < 4; i++) {
            int newX = x + dir[i][0];
            int newY = y + dir[i][1];

            if (isValid(newX, newY, grid.size(), grid[0].size()) &&
                grid[newX][newY] >= grid[x][y]) {
                dfs(newX, newY, grid, vis);
            }
        }
    }

    vector<vector<int>> pacificAtlantic(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<vector<bool>> vis1(m, vector<bool>(n, false));
        vector<vector<bool>> vis2(m, vector<bool>(n, false));

        for (int i = 0; i < m; i++) {
            dfs(i, 0, grid, vis1);
        }

        for (int i = 0; i < n; i++) {
            dfs(0, i, grid, vis1);
        }

        for (int i = 0; i < m; i++) {
            dfs(i, n - 1, grid, vis2);
        }

        for (int i = 0; i < n; i++) {
            dfs(m - 1, i, grid, vis2);
        }

        vector<vector<int>> ans;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (vis1[i][j] && vis2[i][j]) {
                    ans.push_back({i, j});
                }
            }
        }
        return ans;
    }
};